monkey/tests/unit_tests/monkey_island/cc/agent_event_handlers/test_fingerprinting_event_handler.py

```diff
@@ -0,0 +1,202 @@
+    source=AGENT_ID, target=IPv4Address(TARGET_MACHINE_IP), discovered_services=DISCOVERED_SERVICES
+)
+
+FINGERPRINTING_EVENT = FingerprintingEvent(
+    source=AGENT_ID,
+    target=IPv4Address(TARGET_MACHINE_IP),
+    os=OperatingSystem.WINDOWS,
+    os_version="XP",
+    discovered_services=DISCOVERED_SERVICES,
+)
+
+
+@pytest.fixture
+def target_machine() -> Machine:
+    return Machine(
+        id=TARGET_MACHINE_ID,
+        hardware_id=9,
+        network_interfaces=[IPv4Interface(f"{TARGET_MACHINE_IP}/24")],
+    )
+
+
+@pytest.fixture
+def agent_repository() -> IAgentRepository:
+    return MagicMock(spec=IAgentRepository)
+
+
+@pytest.fixture
+def machine_repository(
+    machine_from_id: Callable[[MachineID], Machine],
+    machines_from_ip: Callable[[IPv4Address], Sequence[Machine]],
+) -> IMachineRepository:
+    machine_repository = MagicMock(spec=IMachineRepository)
+    machine_repository.get_machine_by_id = MagicMock(side_effect=machine_from_id)
+    machine_repository.get_machines_by_ip = MagicMock(side_effect=machines_from_ip)
+    machine_repository.get_new_id = MagicMock(side_effect=count(SEED_ID))
+    return machine_repository
+
+
+@pytest.fixture
+def node_repository() -> INodeRepository:
+    return MagicMock(ispec=INodeRepository)
+
+
+@pytest.fixture
+def agent_machine_facade(agent_repository, machine_repository):
+    return AgentMachineFacade(agent_repository, machine_repository)
+
+
+@pytest.fixture
+def network_model_update_facade(agent_machine_facade, machine_repository, node_repository):
+    return NetworkModelUpdateFacade(agent_machine_facade, machine_repository, node_repository)
+
+
+@pytest.fixture
+def fingerprinting_event_handler(network_model_update_facade, machine_repository):
+    return FingerprintingEventHandler(network_model_update_facade, machine_repository)
+
+
+@pytest.fixture
+def machines_by_id(target_machine: Machine) -> Dict[MachineID, Machine]:
+    return {target_machine.id: target_machine}
+
+
+@pytest.fixture
+def machines_by_ip(target_machine: Machine) -> Dict[IPv4Address, Sequence[Machine]]:
+    return {IPv4Address(TARGET_MACHINE_IP): [target_machine]}
+
+
+@pytest.fixture
+def machine_from_id(machines_by_id) -> Callable[[MachineID], Machine]:
+    def inner(id: int) -> Machine:
+        return machines_by_id[id]
+
+    return inner
+
+
+@pytest.fixture
+def machines_from_ip(machines_by_ip) -> Callable[[IPv4Address], Sequence[Machine]]:
+    def inner(ip: IPv4Address) -> Sequence[Machine]:
+        return machines_by_ip[ip]
+
+    return inner
+
+
+def test_fingerprinting_event_handler__target_machine_does_not_exist(
+    fingerprinting_event_handler, machine_repository: IMachineRepository
+):
+    event = FINGERPRINTING_EVENT_NO_OS_INFO
+    machine_repository.get_machines_by_ip = MagicMock(side_effect=UnknownRecordError)
+
+    fingerprinting_event_handler.handle_fingerprinting_event(event)
+
+    expected_machine = Machine(id=SEED_ID, network_interfaces=[IPv4Interface(event.target)])
```

Comment:
Can the expected machine just be the target machine?

Reply:
No, the IDs are different. Since the machine doesn't already exist when the event is processed, `network_model_update_facade.get_or_create_target_machine()` creates a new machine with a different ID (`SEED_ID`).

Reply:
We discussed and decided to mock `NetworkModelUpdateFacade`.